### ISSUE TITLE
Everything just stop working

### ISSUE THREAD

**Issue by vellkan** [emmetio/sublime-text-plugin#50]
Today I get update and everything just stop worked :/

1) TAB key didn't work as always, I read about that and change this to Ctrl+E, but it's gonna change all my workflow
2) When I write header.header it's print <header id="header" class=""></header> - why?
3) When I write in CSS file m:0-a-64 I get margin: 0 auto -64px; Why -64px, not only 64px?
4) When I get color shortcut in CSS I can edit all value, with # - why? Before update I was edit only hex value, after # sign.
4) I lost all my snippets shortcuts for CSS
5) Plugin for generate lipsum text stops working

For now I need uninstall everything and try to re-configure everything again.

I now You've done great job, but for me - it ruined my day and all my workflow. Seriously, I just want to back to earlier version...

**Comment by sergeche** [comment]
> Emmet shortcuts work, but when I write php and hit tab - nothing happen.

Just tried with PHP syntax: wrote `php` then hit Tab and got `<?php  ?>`

> When I write header.header it's print: <header id="header" class="header"></header>

Works fine for me. Maybe you created custom snippet for `header`?

> I found that but I added few own snippets and they don't work.

Can you show me your `Emmet.sublime-settings` file?

> .header {} and write db (for display: block) inside bracket - it won't work.

Fixed in v2.1.2

**Comment by mrcsmcln** [comment]
I'm noticing that Emmet is now super overzealous about tab completion. `php` + Tab turns to `<php></php>` every time.

**Comment by sergeche** [comment]
@mrcsmcln please check your keybindings, did you changed Tab key handler for `emmet_expand_abbreviation` action for previous Emmet version? If so, please remove it

**Comment by mrcsmcln** [comment]
After uninstalling and reinstalling, Emmet works great. Not sure why that needed to happen, as I didn't modify any other settings.

**Comment by sergeche** [comment]
Maybe old version wasn’t correctly unloaded. Anyway, new version should work better with native snippets

**Comment by cheestudio** [comment]
@sergeche I concur, so many things stopped working since this update, mostly all the key bindings. I don't have ANY custom key bindings, they are all default. The biggest ones I use nearly on the hour no longer work:

1) Wrap with Abbreviations (CTRL + SHIFT + G)
2) Jump to Matching Tag (CTRL + ,)

Do I need to define these myself now? Or what else can I check?

**Comment by sergeche** [comment]
@cheestudio Yes, you have to specify your own shortcuts: https://github.com/emmetio/sublime-text-plugin#more-actions

**Comment by cheestudio** [comment]
Sigh. OK. Do you happen to have a default set of those defined that I can copy/paste of the ones you removed? Kind of frustrating to have to map those myself when they've been defined all along...

**Comment by sergeche** [comment]
@cheestudio Lots of people were complaining that Emmet redefined default ST actions or actions from another extension. I'll add sample file with shortcuts later

**Comment by cheestudio** [comment]
Totally, I see why you made the change. Painful in the meantime, but a good change in the long run. But a sample file would be absolutely awesome!

**Comment by rchl** [comment]
The example key bindings could just be included in the default included key bindings configuration, just commented out. That would make it easy to just copy them to user keyboardings and uncomment.

**Comment by vellkan** [comment]
> @cheestudio Lots of people were complaining that Emmet redefined default ST actions or actions from another extension. I'll add sample file with shortcuts later

Yeah, that's true and Your work is amazing. But after this years and stable configuration it's really hard to change habits in one second.

**Comment by sergeche** [comment]
@vellkan it’s not about changing habits, it‘s about spending a few minutes to tweak package for your needs

**Comment by sergeche** [comment]
Check out v2.1.3, there’s new menu item Preferences > Package Settings > Emmet > Key Bindings with shortcuts from previous version, which you can copy and uncomment

**Comment by craigerskine** [comment]
> it’s not about changing habits, it‘s about spending a few minutes to tweak package for your needs

I think it __is__ about changing habits, because TAB no longer works as it did before. I now have to really pay attention to what syntax mode I'm in and use a combination of `TAB`, `CTRL + .`, and `CTRL + E` where I could just use TAB in the previous release.

I added back all the previous key binding snippets but TAB still does not work in `.xsl` files. I used to be able to type:
`xsl:if` + TAB and it would auto complete... now TAB does not work with Emmet in `.xsl` files. I have to use `CTRL + .` or `CTRL + E`. I changed the default `CTRL + E` to just `TAB` and I can partially get it to work like before, but then I can't tab to (un)indent, where I could in the previous release.

So, yes, I do have to change my coding habits. Or is there a way to install the previous version and prevent it from being autoupdated?

**Comment by sergeche** [comment]
> I added back all the previous key binding snippets but TAB still does not work in .xsl files.

It was a bug, fixed it in v2.1.8.

You can mimic Tab key behaviour from previous version with the following keybinding:

```json
{
    "keys": ["tab"],
    "command": "emmet_expand_abbreviation",
    "args": { "tab": true },
    "context": [
        { "key": "emmet_capture_abbreviation" },
        { "key": "selection_empty" },
        { "key": "has_next_field", "operand": false },
        { "key": "auto_complete_visible", "operand": false },
        { "key": "selector", "operand": "source.css, source.sass, source.less, source.scss, source.stylus, source.postcss, source.jade, text.jade, source.pug, text.pug, text.slim, text.xml, text.html - source, text.haml, text.scala.html, source string" }
    ]
}
``` 

But it’s not recommended, please report bugs you struggle with

**Comment by vellkan** [comment]
I appreciate Your work and last updates, but I can't back to previous workflow.
New features are really cool, but I lost many snippets, especially in CSS/SCSS, own snippets didn't work.

I've even try to install ST and plugins from the beginning, but it didn't help.

**Comment by sergeche** [comment]
@vellkan Emmet config format has been changed, everything is available in README. If you have issues with snippets, you can send me your old config and I will help you

**Comment by vellkan** [comment]
> @vellkan Emmet config format has been changed, everything is available in README. If you have issues with snippets, you can send me your old config and I will help you

I know, I've read them. Some of snippets works, some of them - don't work. And everyday I found more things which works differently.

Now I'm testing Visual Studio Code, I'm considering to switch to another app.

**Comment by sergeche** [comment]
@vellkan 
 > I found more things which works differently.

Please report things that doesn't work as expected

**Comment by craigerskine** [comment]
Something I've noticed over the past couple days:
If I open a blank document, change the syntax mode to HTML, none of my custom HTML snippets work anymore. For example, I have a TailwindCSS/Vue starter snippet `tw-html + tab`, when I type that it outputs `<tw-html></tw-html>`

Here's the [Snippet](https://www.dropbox.com/s/4d57zo8wekene7s/Tailwind%20-%20HTML.sublime-snippet?dl=0)

**Comment by sergeche** [comment]
@craigerskine native snippets should not be affected. Maybe you have custom Tab key binding that expands Emmet abbreviation?

**Comment by craigerskine** [comment]
@sergeche Here's my [key bindings](https://github.com/emmetio/sublime-text-plugin/files/5136645/emmet.txt). Is there something wrong with this?

**Comment by sergeche** [comment]
@craigerskine Oh, got it: Emmet detected that `tw-html` looks like Web Component name so it expands it. There are several options:

1. Remove dash from your snippet (e.g. `twhtml`)
2. After typing `tw-html` press Esc to clear Emmet abbreviation then Tab to expand native snippet
3. I can add option to disable CamelCased and kebab-case words for abbreviation capturing

**Comment by craigerskine** [comment]
@sergeche AWESOME! Thanks for your help.

**Comment by vellkan** [comment]
1. Many snippets stops working.
For example:
`bdn` was: border: none;
`ti-` was text-indent: -9999px;
`bg+` was background: url() left top no-repeat;

It's just a few of them. I remove all my CSS/SCSS plugins with snippets, reload them, nothing helps.

2. tag-preview: true doesn't work, I can't get this preview even on clean install only with Emmet

3. When I'm typing something long, for example:
`.header>.row>ul.main>li*4>lorem` app does not recognise lorem and disable TAB action. I need to write partial code, I can't just hit TAB without underline text, because that did't call Emmet action.

**Comment by kirkegaard** [comment]
Im seeing the same issues as above. Im using emmet 2.1.14 with JavaScript (Babel). I can get the expansion to work with a `<` prefix but my `ctrl+e` doesnt work anymore. I havent added anything in my configuration. Im using the plugin with standard configuration.

**Comment by sergeche** [comment]
@kirkegaard you should add key binding manually, see https://github.com/emmetio/sublime-text-plugin#more-actions
Ctrl-e overrides default Go to End of Line action

**Comment by sergeche** [comment]
@vellkan 
> Many snippets stops working.
> For example:
> bdn was: border: none;

Works for me, please check again with latest version

> ti- was text-indent: -9999px;
> bg+ was background: url() left top no-repeat;

All snippets with operators like `-` or `+` are prohibited now: it’s very hard to parse them properly (e.g. ensure that `+` should be a part of the snippet, not a chining operator). There’s no sane reason to use them in snippets, they can be easily replaced with other characters

> It's just a few of them. I remove all my CSS/SCSS plugins with snippets, reload them, nothing helps.

Please provide more examples of snippets that doesn’t work

> tag-preview: true doesn't work, I can't get this preview even on clean install only with Emmet

Should work with HTML syntaxes. If it still doesn’t please restart ST, open ST console and send me it’s contents

> When I'm typing something long, for example:
> .header>.row>ul.main>li*4>lorem app does not recognise lorem and disable TAB action. I need to write partial code, I can't > just hit TAB without underline text, because that did't call Emmet action.

Works for me: 
<img width="464" alt="image" src="https://user-images.githubusercontent.com/93595/91656843-c8b4c080-eac4-11ea-8951-17d83e30a9c6.png">

Please ensure you have the most recent Emmet version

> When I'm typing : I get :;

It’s a default ST behaviour

**Comment by guimadaleno** [comment]
Hi @sergeche. 
Thank you for your effort on this amazing project and as well as your patience with the community.

I've been having some issues with the latest **2.1.16** as well. 
I'm still unsure if it's something with my ST3 settings or a possible Emmet bug.
So I just wanted to show, in practice, what's going on with my ST3. 

![Emmet bug](https://s8.gifyu.com/images/Emmet-Bugs.gif)

If that's a possible bug, hope this could help.

_SublimeText 3.2.2 Build 3211
Emmet 2.1.16 with empty settings
Packages I'm using: Emmet, Hasher, Material Theme, Open Folder, Package Control, PyV8, Sublime Git_

**Comment by sergeche** [comment]
@guimadaleno thanks for the report, looks like a bug. Could you create a separate issue so we can discuss it there?

**Comment by gladisihor** [comment]
Hi @sergeche !
Please help me to deal with my CSS/HTML snippets.

Before Emmet update:
`"bd1" => "border:1px solid #|;"
"c1" => "color:#111;"
"cf" => "color:#fff;"
"lh20" => "line-height:20px;"
"after" => "display:block;\ncontent:\"\";\nclear:both;" `

After Emmet update:
`"bd1" => "border: 1px;"
"c1" => "color: 1px;"
"cf" => NOTHING
"lh20" => "line-height: 20;
"after" => NOTHING `

... and a lot of such examples I've got. Also I can't find how to manage css.valueSeparator option. Previously it was "css.valueSeparator": ":" (without spaces after ":"), but now it doesn't work as well. Another question is how to create HTML snippet with caret position after abbreviation expanding? Sequence of caret position or highlight some code after expanding? For example:
`"js": "<script src=\"js/${1:main.js}\">${2}</script>${3}"`

My old snippets and preferences:
[emmet-st3.zip](https://github.com/emmetio/sublime-text-plugin/files/5152535/emmet-st3.zip)

Thank you in advance :)

**Comment by vellkan** [comment]
@sergeche Thanks for respond - now I understand why some behaviors didn't work like I expect.
I will find new snippets to fill that way.

But I've got same problem like @guimadaleno on his gif. I've try to describe this before, but visualisation is better :)

**Comment by vfonic** [comment]
I'm probably also in this camp. Ever since Emmet upgraded automatically, so many things broke in my ST3.

I'm using TabNine AI-based code completion: https://www.tabnine.com/

Now it's not working well anymore. Emmet and TabNine are fighting for priority and I never know which one will complete when I press <Tab>. Many other keyboard shortcuts are also broken and I'm not sure if this is related or not.

Is there any way to revert, at least temporarily, to the previous verison?

Thanks!

**Comment by sergeche** [comment]
@vfonic by default, Emmet now expands captured abbreviations only (they are underlined): https://github.com/emmetio/sublime-text-plugin#expanding-abbreviation
You can clear abbreviation with Esc key so Tab will be handled by other plugin.

And please, create new issue with detailed description where it’s not working anymore
